### PR TITLE
Fix fmapi_chat for instruct models and custom tokenizers

### PR DESCRIPTION
**Pull request by aspfohl**
🙃 https://stackoverflow.com/questions/67412925/what-is-the-difference-between-lentokenizer-and-tokenizer-vocab-size

Tested via:
```yaml
device_eval_batch_size: 4
icl_tasks:
-
  label: jeopardy
  dataset_uri: eval/local_data/world_knowledge/jeopardy_all.jsonl
  num_fewshot: [3]
  icl_task_type: language_modeling
  continuation_delimiter: "\nAnswer: " # this separates questions from answers
  has_categories: true
icl_subset_num_batches: 1
max_seq_len: 1024
models:
- model:
    local: true
    name: fmapi_chat
  model_name: mpt-example
  tokenizer:
    name: mosaicml/mpt-7b-8k-instruct
seed: 1
```

### CODE REVIEW

File: llmfoundry/models/inference_api_wrapper/openai_causal_lm.py [hunk 1]
```diff
@@ -228,9 +228,14 @@ def process_result(self, completion: Optional['ChatCompletion']):
             tensors = []
             for t in self.tokenizer(
                     completion.choices[0].message.content)['input_ids']:
-                tensors.append(
-                    self.tokenizer.construct_logit_tensor(
-                        {self.tokenizer.decode([t]): 0.0}))
+
+                # Construct tensor of shape (vocab_size,) with logprobs for each token
```
Comment (aspfohl) [inline llmfoundry/models/inference_api_wrapper/openai_causal_lm.py:232]: From https://github.com/mosaicml/llm-foundry/blob/ac7835415cf6fa91664e513c7a471f2802cdc8a2/llmfoundry/tokenizers/tiktoken.py#L361

Comment (dakinggg) [reply llmfoundry/models/inference_api_wrapper/openai_causal_lm.py:232]: Should this fix just be in `tiktoken.py`? or why are you copying it over here?

Comment (dakinggg) [reply llmfoundry/models/inference_api_wrapper/openai_causal_lm.py:232]: nvm

**Review by dakinggg — COMMENTED** [review]
Want to add a test that would've caught your issues?

**Comment by aspfohl** [comment]
@dakinggg  Good call, caught an issue in interface.py that would have caused the same bug when I got to casual models.. 

Updated [8a7c929](https://github.com/mosaicml/llm-foundry/pull/914/commits/8a7c929c17d70ba54152fc90b0be7566bcf1d2b2)

**Review by dakinggg — APPROVED** [review]
Would be nice if @maxisawesome or @bmosaicml also took a glance

File: llmfoundry/models/inference_api_wrapper/openai_causal_lm.py [hunk 2]
```diff
@@ -228,9 +228,15 @@ def process_result(self, completion: Optional['ChatCompletion']):
...
+                tokenizer_logprobs = {self.tokenizer.decode([t]): 0.0}
```
Comment (maxisawesome) [inline llmfoundry/models/inference_api_wrapper/openai_causal_lm.py:233]: Do we want to make this 1.0?

File: llmfoundry/models/inference_api_wrapper/openai_causal_lm.py [hunk 3]
```diff
@@ -228,9 +228,15 @@ def process_result(self, completion: Optional['ChatCompletion']):
...
+                tensor = torch.tensor([min(tokenizer_logprobs.values()) - 1] *
```
Comment (maxisawesome) [inline llmfoundry/models/inference_api_wrapper/openai_causal_lm.py:234]: I know all this is taken from tiktoken.py but this logic doesn't make sense to me.
L233: set each log prob to 0 in a dict
L234/235: take minimum of all tokenizer_logprobs (we set them all to 0) and subtract 1 to get -1, duplicate that value (always -1?) to make a tensor of size len(tokenizer). 
L236-238: at tensor location for each token in tokenizer_logprobs, set probability to 0.0 instead of -1.

So we're not returning tru logprobs at all if I'm not mistaken?

Comment (maxisawesome) [reply llmfoundry/models/inference_api_wrapper/openai_causal_lm.py:234]: @bmosaicml

Comment (aspfohl) [reply llmfoundry/models/inference_api_wrapper/openai_causal_lm.py]: [Silliest refactor ever](https://github.com/mosaicml/llm-foundry/pull/914/commits/861a2bd66ee2cc1c79d1e38c11b32d3de7b5714b), but this should make this significantly clearly

Comment (bmosaicml) [reply llmfoundry/models/inference_api_wrapper/openai_causal_lm.py]: OpenAI only returns top 5 logprobs, but there's still `len(tokenizer) - 5 logprobs` that need to be mocked in order for the metric's math to be computed properly. n.b. recall that the metrics expect to be able to index into a tensor of size len(tokenizer) and inspect the logprob there

**Comment by bmosaicml** [comment]
Great catch! I can only imagine how hard this bug was to track down. 🙃

**Review by bmosaicml — APPROVED** [review]
